raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -498,8 +515,8 @@ private void handleFetchQuorumRecordsResponse(FetchQuorumRecordsResponseData res
         electionTimer.reset(electionTimeoutMs);
     }
 
-    private OptionalLong maybeAppendAsLeader(LeaderState state, Records records) {
-        if (state.highWatermark().isPresent() && stateMachine.accept(records)) {
+    private OptionalLong maybeAppendAsLeader(LeaderState state, Records records, boolean isLeaderChangeMessage) {
+        if ((state.highWatermark().isPresent() || isLeaderChangeMessage) && stateMachine.accept(records)) {
```

Comment:
I'm a bit confused by the first condition. The purpose of the high watermark check (which probably should have a comment) is to ensure that the state machine has seen all of the committed data prior to receiving a new call to `accept`. Also, if it is a control record, do we want to send it to the state machine? I don't think it makes sense for the state machine to be able to reject our control record append with `accept`, but it might be simpler if we still forward the records to the state machine in `apply`. That would prevent us from getting into the same situation with streams where the latest position in the state machine is off by one. So maybe we should have a separate `appendControlRecord` call?